core/Pos/Binary/Core/Blockchain.hs

```diff
@@ -13,32 +13,29 @@ import           Pos.Binary.Core.Block ()
 import           Pos.Binary.Core.Common ()
 import qualified Pos.Core.Block.Blockchain as T
 import           Pos.Core.Block.Union.Types (BlockHeader (..))
-import           Pos.Core.Configuration (HasConfiguration)
-import           Pos.Crypto.Configuration (HasCryptoConfiguration, getProtocolMagic, protocolMagic)
+import           Pos.Crypto.Configuration (ProtocolMagic (..))
 import           Pos.Util.Util (cborError)
 
 instance ( Typeable b
          , Bi (T.BHeaderHash b)
          , Bi (T.BodyProof b)
          , Bi (T.ConsensusData b)
          , Bi (T.ExtraHeaderData b)
-         , HasCryptoConfiguration
          ) =>
          Bi (T.GenericBlockHeader b) where
     encode bh =  encodeListLen 5
-              <> encode (getProtocolMagic protocolMagic)
+              <> encode (getProtocolMagic (T._gbhProtocolMagic bh))
               <> encode (T._gbhPrevBlock bh)
               <> encode (T._gbhBodyProof bh)
               <> encode (T._gbhConsensus bh)
               <> encode (T._gbhExtra bh)
     decode = do
         enforceSize "GenericBlockHeader b" 5
-        blockMagic <- decode
-        -- TODO include ProtocolMagic in the definition of GenericBlockHeader,
-        -- and decode it, eliminating this failure case. Protocol magic checks
-        -- must not happen in decoding.
+        _gbhProtocolMagic <- ProtocolMagic <$> decode
+        {- TODO replace this in block verification.
         when (blockMagic /= getProtocolMagic protocolMagic) $ cborError $
```

Comment:
This looks important.

Reply:
!! Indeed. Working on it now

Reply:
I put it into `verifyHeader` (not pushed yet), but it's not clear to me whether block verification even uses `verifyHeader`.

Reply:
@avieth it does: `Pos.Block.Logic.Integrity.verifyHeader` -> `Pos.Block.Logic.Integrity.verifyBlocks` -> `slogVerifyBlocks` -> `verifyBlocksPrefix` -> ..., and also it is called by `classifyNewHeader` (called by `handleUnsolicitedHeader`)

`->` = `called by`

Reply:
Nice, thanks @coot